### PR TITLE
Add legacy mode to Packer

### PR DESCRIPTION
**Pull request by nywilken**
<details>
<summary> Add concept of Legacy mode to Packer</summary>

In the legacy mode users running build or validate with JSON templates will get a warning indicating that Packer is in Legacy Configuration Mode.

Legacy mode can be controlled by an environment variable.

PACKER_LEGACY_MODE=      // default if not set will display warnings to the user.
PACKER_LEGACY_MODE=off // will display warnings to the user when running, and may disable other features.
PACKER_LEGACY_MODE=on // will not display warnings with all legacy features enabled.

    ~>  go run . validate ../packer-templates/amazon-chroot/amazon-chroot_shell.json

    Legacy JSON Configuration Mode enabled.
    The template will be parsed in the legacy configuration style. Legacy style
    templates will continue to work but users are encouraged to move HCL style templates.
    templates will continue to work but users are encouraged to move to HCL style templates.
    See: https://learn.hashicorp.com/tutorials/packer/hcl2-upgrade

    The configuration is valid.

    ~>  PACKER_LEGACY_MODE=on go run . validate ../packer-templates/amazon-chroot/amazon-chroot_shell.json
    The configuration is valid.

</details>

<details>
<summary>Disable vendored plugins when PACKER_LEGACY_MODE is off</summary>

This change allows users to disable the loading of vendored plugins by setting legacy mode off. In this mode Packer will rely only on plugins that have been installed via packer init or the packer plugins command.

    // Running validate on JSON with Packer legacy mode unset; all plugins loaded by default
    ~>  go run . validate ../packer-templates/amazon-ebs/amazon-ebs_ubuntu_shell_manifest.json

    Legacy JSON Configuration Mode enabled.
    The template will be parsed in the legacy configuration style. Legacy style
    templates will continue to work but users are encouraged to move HCL style templates.
    templates will continue to work but users are encouraged to move to HCL style templates.
    See: https://learn.hashicorp.com/tutorials/packer/hcl2-upgrade

    In JSON mode automatic loading of vendored plugins is enabled by default.  In future
    this feature will be removed and Packer will rely only on system installed plugins.
    To disable automatic loading of vendored plugins set PACKER_LEGACY_MODE=off
    See: https://packer.io/docs/templates/legacy-mode

    The configuration is valid.

    // Running validate on JSON with Packer legacy mode explicitly set to on; all plugins loded
    ~>  PACKER_LEGACY_MODE=on go run . validate ../packer-templates/amazon-ebs/amazon-ebs_ubuntu_shell_manifest.json
    The configuration is valid.

    // Running validate on JSON with Packer legacy mode explicitly set to off
    ~>  PACKER_LEGACY_MODE=off go run . validate ../packer-templates/amazon-ebs/amazon-ebs_ubuntu_shell_manifest.json

    Legacy JSON Configuration Mode enabled.
    The template will be parsed in the legacy configuration style. Legacy style
    templates will continue to work but users are encouraged to move HCL style templates.
    templates will continue to work but users are encouraged to move to HCL style templates.
    See: https://learn.hashicorp.com/tutorials/packer/hcl2-upgrade

    In JSON mode automatic loading of vendored plugins is enabled by default.  In future
    this feature will be removed and Packer will rely only on system installed plugins.
    To disable automatic loading of vendored plugins set PACKER_LEGACY_MODE=off
    See: https://packer.io/docs/templates/legacy-mode

    Error: Failed to initialize build "amazon-ebs"

    error initializing builder 'amazon-ebs': Unknown builder amazon-ebs

    exit status 1
    
</details>

See related PRs:
- #11562 
- #11563 

## TODO

- [ ] Tests
- [ ] Docs

Closes https://github.com/hashicorp/packer-internal-issues/issues/16 (HashiCorp-only link)

### CODE REVIEW

**Review by azr — COMMENTED** [review]
Initial review, looks pretty nice !

File: hcl2template/types.required_plugins.go [hunk 1]
```diff
@@ -99,19 +98,6 @@ func (cfg *PackerConfig) decodeImplicitRequiredPluginsBlock(k ComponentKind, blo
 	// this makes this simple.
 	componentName := block.Labels[0]
 
-	store := map[ComponentKind]packer.BasicStore{
-		Builder:       cfg.parser.PluginConfig.Builders,
-		PostProcessor: cfg.parser.PluginConfig.PostProcessors,
-		Provisioner:   cfg.parser.PluginConfig.Provisioners,
-		Datasource:    cfg.parser.PluginConfig.DataSources,
-	}[k]
-	if store.Has(componentName) {
```
[the remark this answers is not in the document]
Comment (nywilken) [reply hcl2template/types.required_plugins.go:108]: Good questions, it make be easier if I break out each set of changes into their own PR to help here. This PR contains a number of changes to make various Packer features work in Legacy mode. I described each use case in the description. 

For this particular use case you can see the details under **Update implicitly required plugins to work even in Legacy Mode**

File: command/vendored_plugins.go [hunk 2]
```diff
@@ -187,6 +190,14 @@ var VendoredPostProcessors = map[string]packersdk.PostProcessor{
 // Upon init lets load up any plugins that were vendored manually into the default
 // set of plugins.
 func init() {
+
+	// By default PACKER_LEGACY_MODE=auto to enable automatic loading of vendored plugins.
+	// In preparation for removing vendored plugins entirely we are providing users a flag to
+	//  to control if vendored plugins should be loaded or not.
+	if v, ok := os.LookupEnv("PACKER_LEGACY_MODE"); ok && strings.ToLower(v) == "off" {
```
[the remark this answers is not in the document]
Comment (nywilken) [reply command/vendored_plugins.go:197]: Good call, I'll update. Previously I was not checking the value so I went with LookupEnv. But Getenv is much better now. Thnx

File: command/build.go [hunk 3]
```diff
@@ -147,7 +172,11 @@ func (m *Meta) GetConfigFromJSON(cla *MetaArgs) (packer.Handler, int) {
 func (c *BuildCommand) RunContext(buildCtx context.Context, cla *BuildArgs) int {
 	packerStarter, ret := c.GetConfig(&cla.MetaArgs)
 	if ret != 0 {
-		return ret
+		return 1
```
[the remark this answers is not in the document]
Comment (nywilken) [reply command/build.go:175]: I think that is an oops 😄 I'll revert that.

File: hcl2template/types.required_plugins.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nywilken) [reply hcl2template/types.required_plugins.go:108]: That said, I will add a test for the use case you just mentioned and see how things behave.

File: main.go [hunk 4]
```diff
@@ -326,67 +326,67 @@ func loadConfig() (*config, error) {
...
-			//"amazon-chroot":       "github.com/hashicorp/amazon",
-			//"amazon-ebs":          "github.com/hashicorp/amazon",
-			//"amazon-ebssurrogate": "github.com/hashicorp/amazon",
-			//"amazon-ebsvolume":    "github.com/hashicorp/amazon",
-			//"amazon-instance":     "github.com/hashicorp/amazon",
+			"amazon-chroot":       "github.com/hashicorp/amazon",
+			"amazon-ebs":          "github.com/hashicorp/amazon",
+			"amazon-ebssurrogate": "github.com/hashicorp/amazon",
+			"amazon-ebsvolume":    "github.com/hashicorp/amazon",
+			"amazon-instance":     "github.com/hashicorp/amazon",
```
[the remark this answers is not in the document]
Comment (nywilken) [reply main.go:333]: This was for testing implicit plugins in Legacy mode. Each of these changes are their own commits with tests runs in the commit message and PR description to help explain the use case. I will break this into its own PR

File: command/hcl2_upgrade.go [hunk 5]
```diff
@@ -159,9 +159,6 @@ func (c *HCL2UpgradeCommand) RunContext(_ context.Context, cla *HCL2UpgradeArgs)
 	}
 
 	core := hdl.(*CoreWrapper).Core
-	if err := core.Initialize(); err != nil {
-		c.Ui.Error(fmt.Sprintf("Ignoring following initialization error: %v", err))
-	}
```
[the remark this answers is not in the document]
Comment (nywilken) [reply command/hcl2_upgrade.go:164]: Yeah I agree with your comment "It might also have interesting errors when the plugin is actually installed." So I went back and forth on it. I didn't read over the documentation on these errors for hcl2_upgrade but it did feel a little weird to have an error, followed by a successful file creation. 

I added some alternative text to suggest the user install missing builders before running. Since we are ignoring maybe rewording a little will make it is less weird for me. I'll revisit here.

[the remark this answers is not in the document]
Comment (azr) [reply command/hcl2_upgrade.go:164]: One thing we could do too is manually ignore the 'component not found' errors.

File: hcl2template/types.required_plugins.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nywilken) [reply hcl2template/types.required_plugins.go:108]: haha - I wanted to tell you that you did a great job with the plugin requirements check. I was very curious about this as well. I found that it does now because the call ListInstallationsOptions does a good of detecting that a plugin has already been installed so no need to implicitly install. 

```
// Running init against template with required plugin block for github.com/hashicorp/amazon v1.0.1
~>  PACKER_LEGACY_MODE=on go run . init ../packer-templates/amazon-ebs/amazon-ebs_ubuntu_shell-local.pkr.hcl
Installed plugin github.com/hashicorp/amazon v1.0.1 in "~/.packer.d/plugins/github.com/hashicorp/amazon/packer-plugin-amazon_v1.0.1_x5.0_darwin_amd64"

// Running init against template with no required plugin block after a successful init for github.com/hashicorp/amazon v1.0.1
~>  PACKER_LEGACY_MODE=on go run . init ../packer-templates/amazon-ebs/amazon-ebs_ubuntu_shell-local.pkr.hcl
```

[the remark this answers is not in the document]
Comment (azr) [reply hcl2template/types.required_plugins.go]: nice !

**Review by azr — APPROVED** [review]
Nice one, LGTM !

File: command/build.go [hunk 6]
```diff
@@ -97,22 +98,46 @@ func (m *Meta) GetConfig(cla *MetaArgs) (packer.Handler, int) {
...
+
+	coloredUi := packer.ColoredUi{
+		Color: packer.UiColorRed,
+		Ui:    ui,
+	}
+
+	coloredUi.Say(`
+Legacy JSON Configuration Mode enabled.
+The template will be parsed in the legacy configuration style. Legacy style
+templates will continue to work but users are encouraged to move HCL style templates.
```
Comment (sylviamoss) [inline command/build.go:128]: Do we plan to remove JSON support entirely? If we do, maybe we can say that will continue to work for a limited time

File: command/build.go [hunk 7]
```diff
@@ -147,7 +172,11 @@ func (m *Meta) GetConfigFromJSON(cla *MetaArgs) (packer.Handler, int) {
 func (c *BuildCommand) RunContext(buildCtx context.Context, cla *BuildArgs) int {
 	packerStarter, ret := c.GetConfig(&cla.MetaArgs)
 	if ret != 0 {
-		return ret
+		return 1
+	}
+
+	if cla.ConfigType == ConfigTypeJSON {
+		DisplayLegacyConfigWarning(c.Ui)
```
[the remark this answers is not in the document]
Comment (sylviamoss) [reply command/build.go:179]: +1

**Review by sylviamoss — APPROVED** [review]
Nice! This looks good 👍🏼
